### ISSUE TITLE
app router does not resolve routes according to `generateStaticParams` constraints

### ISSUE THREAD

**Issue by y-nk** [vercel/next.js#69770]
### Link to the code that reproduces this issue

https://stackblitz.com/edit/stackblitz-starters-941nyd

### To Reproduce

1. click on "post 2" → it's a post
1. click on "post 1" → it's a post
1. click on "page 2" → **it's a post**
1. click on "page 1" → **it's a post**

### Current vs. Expected behavior

the "pages" urls should resolve with the page route, not the post route.

### Provide environment information

```bash
irrelevant, happens in stackblitz too.
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local)

**Comment by y-nk** [comment]
somehow related: https://github.com/vercel/next.js/issues/53887

**Comment by y-nk** [comment]
looking further, the regexes generated for the routes seem to ignore totally the `generateStaticParams`

![image](https://github.com/user-attachments/assets/99b85441-c162-48f6-9845-9fdae8ac31f8)

**Comment by y-nk** [comment]
after a good round of `console.log` (and from my candid PoV of next), it seems like the `AppPageRouteMatcher` does not contain data from `getStaticParams`, therefore it's impossible to get an accurate route matching.

Route definition is passed here, but lacks of any kind of ref to `getStaticParams` :/

https://github.com/vercel/next.js/blob/a5f30e68f8d1fe614965abc5df0234a869c6db8a/packages/next/src/server/route-matcher-providers/app-page-route-matcher-provider.ts#L43-L45

**Comment by y-nk** [comment]
of course i want to state clear that i know this:

![image](https://github.com/user-attachments/assets/75a3c5f8-3c8a-49db-a2ec-6bf9cf4ad1dc)

but still routes imho should abide to the result of this function and pass to other matchers before making a decision. here is seems the logic is → match the path blindly regardless of the values in params, then render the route no matter what.

it's utterly deceptive because it's like not using a data validation library (the valid values returned from `generateStaticParams`) _and then_ pass unvalidated/wrong values to a component.

**Comment by y-nk** [comment]
@abhi12299 i have found an (slightly ugly) way to make it work, which i think could be used as a "ultimate manual mode" feature.

currently my idea is to add some code for manual/force route matching based on headers here:

https://github.com/vercel/next.js/blob/a5f30e68f8d1fe614965abc5df0234a869c6db8a/packages/next/src/server/lib/router-utils/resolve-routes.ts#L264-L272

such as:

```ts
            for (const route of dynamicRoutes) {
              if (req.headers['x-resolved-route']) {
                // skip the desired route was provided in headers
                if (route.page !== req.headers['x-resolved-route']) {
                  continue;
                }
              }
```

then we can add this internal header in the middleware like this:

![image](https://github.com/user-attachments/assets/a9131f11-8347-4733-bdbd-a001ace3497a)

**Comment by abhi12299** [comment]
This is equivalent to hard coding the route match rules in the codebase.

**Comment by y-nk** [comment]
yes it is. but i doubt fixing the dev server will be easier. this is a 3 lines patch which can be applied with patch-package and gives the hand to userland through the middleware. so i think it's pretty ok until the maintainers finally decide to have a look at this. (cc @leerob i had mentioned this to you on X)

**Comment by abhi12299** [comment]
yeah - its' not a trivial fix. let's see what happens with this

**Comment by marcoshernanz** [comment]
Thanks for the report. I reproduced this on 16.3.4 and 16.4.0-canary.15. Both route patterns match every URL with at least two segments. `generateStaticParams` controls prerendering, not route selection, and `dynamicParams` defaults to `true`, so unlisted params remain valid. Production happens to prefer exact prerendered entries, but that should not be used to disambiguate competing routes. Please use non-overlapping route shapes or a single catch-all with explicit validation. Closing as expected behavior. Related production collision handling is tracked in #81558.

**Comment by y-nk** [comment]
@marcoshernanz thanks for looking into this again (it's been a while but late is better than never). don't you think it's deceptive that despite writing a generateStaticParams function, its result is ignored in dev mode?

Reopen: i disagree (or at least _do not understand why you say_) that routes are overlapping. regexes are showing that it should match [slug], [locale]/[slug] and if not, [...slug]: [slug] is not overlapping [locale]/[slug] and stating that [...slug] is overlapping would be the same as saying an escape hatch route (*) would overlap any other route and that we should not do that :/
